hail/python/hailtop/batch_client/aioclient.py

```diff
@@ -952,45 +1190,97 @@ async def _submit_job_bunches(self,
-                    start_job_id = await self._commit_update(update_id)
-                    self._submission_info = BatchSubmissionInfo(used_fast_path=False)
-                    assert start_job_id == 1
-                log.info(f'created batch {self.id}')
-            else:
-                if n_job_bunches == 0:
-                    log.warning('Tried to submit an update with 0 jobs. Doing nothing.')
-                    return None
-                if n_job_bunches == 1:
-                    start_job_id = await self._update_fast(byte_job_specs_bunches[0], job_bunch_sizes[0], job_progress_task)
-                else:
-                    update_id = await self._create_update()
-                    await self._submit_job_bunches(update_id, byte_job_specs_bunches, job_bunch_sizes, job_progress_task)
-                    start_job_id = await self._commit_update(update_id)
-                    self._submission_info = BatchSubmissionInfo(used_fast_path=False)
-                log.info(f'updated batch {self.id}')
-            return start_job_id
+                    if n_job_bunches == 0 and n_job_group_bunches == 0:
+                        log.warning('Tried to submit an update with 0 jobs and 0 job groups. Doing nothing.')
+                        return (None, None)
+                    if ((n_job_bunches <= 1 and n_job_group_bunches <= 1) and
+                        ((n_job_bunches == 1 and n_job_group_bunches == 1 and (len(byte_job_specs_bunches[0]) + len(byte_job_group_specs_bunches[0]) <= max_bunch_bytesize)) or
+                            (n_job_bunches == 1 and n_job_group_bunches == 0) or
+                            (n_job_bunches == 0 and n_job_group_bunches == 1))):
+                        start_job_id, start_job_group_id = await self._update_fast(byte_job_specs_bunches[0] if n_job_bunches == 1 else [],
+                                                                                   job_bunch_sizes[0] if n_job_bunches == 1 else 0,
+                                                                                   byte_job_group_specs_bunches[0] if n_job_group_bunches == 1 else [],
+                                                                                   job_group_bunch_sizes[0] if n_job_group_bunches == 1 else 0,
+                                                                                   job_progress_task,
+                                                                                   job_group_progress_task)
+                    else:
+                        start_job_group_id = None
+                        if len(self._job_groups) > 0:
+                            await self._submit_job_group_bunches(byte_job_group_specs_bunches, job_group_bunch_sizes, job_group_progress_task)
+                            # we need to recompute the specs now that the job group ID is absolute
+                            job_specs = [spec.to_dict() for spec in self._job_specs]
```

Comment:
Is there some hidden mutation here that means running this code again produces a different result? That's pretty spooky and tells me that this should be restructured.